## Test 17: Memory Leak Check [1 pts]

```

if ! ( which valgrind &> /dev/null ); then
    # "valgrind is not installed. Please install (as root) with:"
    # "pacman -Sy valgrind"
    test_end 1
fi

leak_output=$(timeout 10 valgrind \
    --trace-children=no \
    --child-silent-after-fork=yes \
    --leak-check=full \
    --track-fds=yes \
    --show-leak-kinds=all \
    --track-origins=yes \
    ./$SHELL_NAME < <(echo "${script}") 2>&1)

echo "${leak_output}"
==2749093== Memcheck, a memory error detector
==2749093== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==2749093== Using Valgrind-3.20.0 and LibVEX; rerun with -h for copyright info
==2749093== Command: ./shshsh
==2749093== 
shshsh.c:145:main(): Input command: ls /
bin
boot
bootstrap.sh
dev
etc
home
lib
lib64
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
shshsh.c:145:main(): Input command: !1
bin
boot
bootstrap.sh
dev
etc
home
lib
lib64
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
shshsh.c:145:main(): Input command: 
shshsh.c:145:main(): Input command: 
shshsh.c:145:main(): Input command: jobs
exec: No such file or directory
shshsh.c:145:main(): Input command: cd
==2749093== Syscall param chdir(path) points to unaddressable byte(s)
==2749093==    at 0x4A942AB: chdir (syscall-template.S:120)
==2749093==    by 0x109CBA: main (shshsh.c:183)
==2749093==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==2749093== 
shshsh.c:145:main(): Input command: ls / # Doing some more lsing
bin
boot
bootstrap.sh
dev
etc
home
lib
lib64
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
shshsh.c:145:main(): Input command: !!
exec: No such file or directory
shshsh.c:145:main(): Input command: asdfghjklqprewopiqwualasdf # Bad Command!
exec: No such file or directory
shshsh.c:145:main(): Input command: # Comment only
shshsh.c:145:main(): Input command: pwd
/home/skuan/first-day/P3-marcus0329
shshsh.c:145:main(): Input command: echo hi
hi
shshsh.c:145:main(): Input command: sort -r -n < /etc/passwd > /tmp/11610
sort: cannot read: '<': No such file or directory
shshsh.c:145:main(): Input command: rm /tmp/11610
rm: cannot remove '/tmp/11610': No such file or directory
shshsh.c:145:main(): Input command: history
shshsh.c:145:main(): Input command: !p
exec: No such file or directory
1 ls /
2 !1
3 
4 
5 jobs
6 cd
7 ls / # Doing some more lsing
8 !!
9 asdfghjklqprewopiqwualasdf # Bad Command!
10 # Comment only
11 pwd
12 echo hi
13 sort -r -n < /etc/passwd > /tmp/11610
14 rm /tmp/11610
15 history
shshsh.c:145:main(): Input command: 
shshsh.c:145:main(): Input command: 
shshsh.c:145:main(): Input command: echo bye
bye
getline: Inappropriate ioctl for device
==2749093== Invalid read of size 1
==2749093==    at 0x4847E54: strcpy (vg_replace_strmem.c:558)
==2749093==    by 0x1099F1: main (shshsh.c:134)
==2749093==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==2749093== 
==2749093== 
==2749093== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==2749093==  Access not within mapped region at address 0x0
==2749093==    at 0x4847E54: strcpy (vg_replace_strmem.c:558)
==2749093==    by 0x1099F1: main (shshsh.c:134)
==2749093==  If you believe this happened as a result of a stack
==2749093==  overflow in your program's main thread (unlikely but
==2749093==  possible), you can try to increase the size of the
==2749093==  main thread stack using the --main-stacksize= flag.
==2749093==  The main thread stack size used in this run was 8388608.
==2749093== 
==2749093== FILE DESCRIPTORS: 9 open (3 std) at exit.
==2749093== Open file descriptor 63:
==2749093==    <inherited from parent>
==2749093== 
==2749093== Open file descriptor 23: /dev/pts/ptmx
==2749093==    <inherited from parent>
==2749093== 
==2749093== Open file descriptor 22: /dev/pts/ptmx
==2749093==    <inherited from parent>
==2749093== 
==2749093== Open file descriptor 21: /home/skuan/.local/share/code-server/logs/20230124T153642/remoteagent.log
==2749093==    <inherited from parent>
==2749093== 
==2749093== Open file descriptor 20: /dev/pts/ptmx
==2749093==    <inherited from parent>
==2749093== 
==2749093== Open file descriptor 19: /home/skuan/.local/share/code-server/logs/20230124T153642/ptyhost.log
==2749093==    <inherited from parent>
==2749093== 
==2749093== 
==2749093== HEAP SUMMARY:
==2749093==     in use at exit: 37,408 bytes in 25 blocks
==2749093==   total heap usage: 64 allocs, 39 frees, 62,304 bytes allocated
==2749093== 
==2749093== 32 bytes in 1 blocks are still reachable in loss record 1 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x109DF6: elist_create (elist.c:28)
==2749093==    by 0x109343: hist_init (history.c:17)
==2749093==    by 0x109827: main (shshsh.c:95)
==2749093== 
==2749093== 256 bytes in 1 blocks are possibly lost in loss record 2 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x109831: main (shshsh.c:100)
==2749093== 
==2749093== 1,048 bytes in 1 blocks are still reachable in loss record 3 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x10987B: main (shshsh.c:109)
==2749093== 
==2749093== 2,280 bytes in 19 blocks are definitely lost in loss record 4 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x4A0EE2E: getdelim (iogetdelim.c:65)
==2749093==    by 0x1096B4: readinput (shshsh.c:29)
==2749093==    by 0x1099DC: main (shshsh.c:134)
==2749093== 
==2749093== 4,096 bytes in 1 blocks are still reachable in loss record 5 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x4A0D470: _IO_file_doallocate (filedoalloc.c:101)
==2749093==    by 0x4A1C56F: _IO_doallocbuf (genops.c:347)
==2749093==    by 0x4A1C56F: _IO_doallocbuf (genops.c:342)
==2749093==    by 0x4A1B2EC: _IO_file_underflow@@GLIBC_2.2.5 (fileops.c:485)
==2749093==    by 0x4A0EEBF: getdelim (iogetdelim.c:77)
==2749093==    by 0x1096B4: readinput (shshsh.c:29)
==2749093==    by 0x1099DC: main (shshsh.c:134)
==2749093== 
==2749093== 4,096 bytes in 1 blocks are still reachable in loss record 6 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x4A0D470: _IO_file_doallocate (filedoalloc.c:101)
==2749093==    by 0x4A1C56F: _IO_doallocbuf (genops.c:347)
==2749093==    by 0x4A1C56F: _IO_doallocbuf (genops.c:342)
==2749093==    by 0x4A1B5D7: _IO_file_overflow@@GLIBC_2.2.5 (fileops.c:744)
==2749093==    by 0x4A1A6ED: _IO_new_file_xsputn (fileops.c:1243)
==2749093==    by 0x4A1A6ED: _IO_file_xsputn@@GLIBC_2.2.5 (fileops.c:1196)
==2749093==    by 0x49ED708: __printf_buffer_flush_to_file (printf_buffer_to_file.c:59)
==2749093==    by 0x49ED7C3: __printf_buffer_to_file_done (printf_buffer_to_file.c:120)
==2749093==    by 0x49F72A9: __vfprintf_internal (vfprintf-internal.c:1460)
==2749093==    by 0x49ECEAE: printf (printf.c:33)
==2749093==    by 0x109518: hist_print (history.c:56)
==2749093==    by 0x109CE8: main (shshsh.c:189)
==2749093== 
==2749093== 25,600 bytes in 1 blocks are still reachable in loss record 7 of 7
==2749093==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==2749093==    by 0x109E4C: elist_create (elist.c:38)
==2749093==    by 0x109343: hist_init (history.c:17)
==2749093==    by 0x109827: main (shshsh.c:95)
==2749093== 
==2749093== LEAK SUMMARY:
==2749093==    definitely lost: 2,280 bytes in 19 blocks
==2749093==    indirectly lost: 0 bytes in 0 blocks
==2749093==      possibly lost: 256 bytes in 1 blocks
==2749093==    still reachable: 34,872 bytes in 5 blocks
==2749093==         suppressed: 0 bytes in 0 blocks
==2749093== 
==2749093== For lists of detected and suppressed errors, rerun with: -s
==2749093== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
timeout: the monitored command dumped core

# Check for open FDs
awk "${fd_check}" <<< "${leak_output}" \
    | grep -i '==[0-9]*==.*file descriptor' && test_end 1

# Make sure there were no leaks possible
grep -i '==[0-9]*==.*no leaks are possible' \
    <<< "${leak_output}" || test_end 1
 --> Test failed (1)
```

